hw08  Автоматическое логирование.

Taisiya-Boon opened the pull request:
Разработайте такой функционал:
      метод класса можно пометить самодельной аннотацией @Log. При вызове этого метода "автомагически" в консоль должны логироваться значения параметров. Обратите внимание: явного вызова логирования быть не должно.

Taisiya-Boon committed:
hw08  Автоматическое логирование.
      Разработайте такой функционал:
      метод класса можно пометить самодельной аннотацией @Log. При вызове этого метода "автомагически" в консоль должны логироваться значения параметров. Обратите внимание: явного вызова логирования быть не должно.

petrelevich commented on hw08-byte-codes/src/main/java/ru/otus/Agent.java:
```diff
@@ -0,0 +1,165 @@
+            @Override
+            public MethodVisitor visitMethod(int access, String name, String descriptor, String signature, String[] exceptions) {
+                methodsName.add(name);
+                descMethods.put(name, descriptor);
```
как это будет работать с перегруженными методами?

petrelevich commented on hw08-byte-codes/src/main/java/ru/otus/Agent.java:
```diff
@@ -0,0 +1,165 @@
+
+                mv.visitFieldInsn(Opcodes.GETSTATIC, "java/lang/System", "out", "Ljava/io/PrintStream;");
+                mv.visitVarInsn(setOnOpcodes(descMethods.get(name.getKey())), 1);
+                mv.visitInvokeDynamicInsn("makeConcatWithConstants", "(" + setOnType(descMethods.get(name.getKey())) + ")Ljava/lang/String;", handle, "executed method: " + name.getKey() + ", param: \u0001");
```
что если у метода будет несколько параметров?
как они залогируются?

Taisiya-Boon committed:
hw08  Автоматическое логирование.
      работает с перегруженными методами и с методами у которых несколько переменных.

petrelevich commented on hw08-byte-codes/src/main/java/ru/otus/Agent.java:
```diff
@@ -0,0 +1,294 @@
+
+                            mv.visitMethodInsn(Opcodes.INVOKEVIRTUAL, "java/io/PrintStream", "println", "(Ljava/lang/String;)V", false);
+                        } else {
+                            if (i == 1) {
```
идея с этим if не очень понятна.
вторая ветка от третей, вроде бы, совсем не отличается.

Taisiya-Boon replied:
идея состоит в том,  что с первым параметром пишется название метода и этот параметр, со вторым и последующими просто параметр,  а с последним параметром в добавок к написанию параметра происходит перевод курсора консоли на следующую строку. Если параметр не последний вызывается функция print, если последний - println.

Taisiya-Boon committed:
hw08  Автоматическое логирование.
      добавлен комментарий в непонятное место в коде.

Taisiya-Boon committed:
Merge branch 'master' into hw08-byte-codes
